package/MDAnalysis/analysis/bat.py

```diff
@@ -0,0 +1,422 @@
+        self._root_XYZ_inds = [ag_top_inds.index(a.index) for a in self._root]
+        self._torsion_XYZ_inds = [[ag_top_inds.index(a.index) for a in t] \
+          for t in self._torsions]
+
+        # The primary torsion is the first torsion on the list
+        # with the same central atoms
+        prior_atoms = [sorted([a1, a2]) for (a0, a1, a2, a3) in self._torsions]
+        self._primary_torsion_indices = [prior_atoms.index(prior_atoms[n]) \
+          for n in range(len(prior_atoms))]
+
+        self._ag1 = mda.AtomGroup([ag[0] for ag in self._torsions])
+        self._ag2 = mda.AtomGroup([ag[1] for ag in self._torsions])
+        self._ag3 = mda.AtomGroup([ag[2] for ag in self._torsions])
+        self._ag4 = mda.AtomGroup([ag[3] for ag in self._torsions])
+
+    def _prepare(self):
+        self.bat = []
+
+    def _single_frame(self):
+        # Calculate coordinates based on the root atoms
+        # The rotation axis is a normalized vector pointing from atom 0 to 1
+        # It is described in two degrees of freedom
+        # by the polar angle and azimuth
+        (p0, p1, p2) = self._root.positions
+        v01 = p1 - p0
+        v21 = p1 - p2
+        # Internal coordinates
+        r01 = np.sqrt(np.sum(v01 *
+                             v01))  # Distance between first two root atoms
+        r12 = np.sqrt(np.sum(v21 *
+                             v21))  # Distance between second two root atoms
+        a012 = np.arccos(max(-1.,min(1.,np.sum(v01*v21)/\
+          np.sqrt(np.sum(v01*v01)*np.sum(v21*v21))))) # Angle between root atoms
+        # Exernal coordinates
+        e = v01 / r01
+        phi = np.arctan2(e[1], e[0])  # Polar angle
+        theta = np.arccos(e[2])  # Azimuthal angle
+        # Rotation to the z axis
+        cp = np.cos(phi)
+        sp = np.sin(phi)
+        ct = np.cos(theta)
+        st = np.sin(theta)
+        Rz = np.array([[cp * ct, ct * sp, -st], [-sp, cp, 0],
+                       [cp * st, sp * st, ct]])
+        pos2 = Rz.dot(p2 - p1)
+        # Angle about the rotation axis
+        omega = np.arctan2(pos2[1], pos2[0])
+        root_based = np.concatenate((p0, [phi, theta, omega, r01, r12, a012]))
+
+        # Calculate internal coordinates from the torsion list
+        bonds = calc_bonds(self._ag1.positions,
+                           self._ag2.positions,
+                           box=self._ag1.dimensions)
+        angles = calc_angles(self._ag1.positions,
+                             self._ag2.positions,
+                             self._ag3.positions,
+                             box=self._ag1.dimensions)
+        torsions = calc_dihedrals(self._ag1.positions,
+                                  self._ag2.positions,
+                                  self._ag3.positions,
+                                  self._ag4.positions,
+                                  box=self._ag1.dimensions)
+        # When appropriate, calculate improper torsions
+        torsions = np.array([\
```

Comment:
I think there's probably a faster way to do this with numpy masking rather than a list comprehension.  Things like this are going to get slow for big systems.

Reply:
I've rewritten these lines to be based on numpy arrays instead of a list comprehension